app/components/detail-topics.js

```diff
@@ -1,28 +1,85 @@
 import Ember from 'ember';
+import DS from 'ember-data';
+
+const {computed, RSVP} = Ember;
+const {PromiseArray} = DS;
 
 export default Ember.Component.extend({
   classNames: ['detail-topics'],
   subject: null,
   isManaging: false,
-  //keep track of our initial state so we can roll back
-  initialTopics: [],
+  isSaving: false,
+  isCourse: false,
+  isSession: false,
+  isProgramYear: false,
+  bufferedTopics: [],
+  availableTopics: computed('isCourse', 'isSession', 'isProgramYear', 'subject', function(){
+    let defer = RSVP.defer();
+    
+    if(this.get('isCourse')){
+      this.get('subject').get('school').then(school => {
+        school.get('topics').then(topics => {
+          defer.resolve(topics);
+        });
+      });
+    }
+    
+    if(this.get('isSession')){
```

Comment:
Yea. I'm not sure how to do this right. Ideally I would probably create subclass components for each of these states but that seems like over kill. I felt like else if made it harder to understand because the blocks seemed related. Maybe I should stick a return inside each block so the next one isn't reached?  Or move the sets out into functions here?  Or I can just go else if, but it seems harder to read that way.